### PR TITLE
fix: support pluginManagement in mvn.pom preprocessor

### PR DESCRIPTION
**Pull request by danielmoy-google**
Support both ways:

1. Parent pom.xml files which specify `maven-compiler-plugin` inside of
   `pluginManagement/plugins/plugin` instead of `plugins/plugin` should
   be correctly handled, and verified.
2. Child pom.xml files which specify a bare `maven-compiler-plugin` that
   references a parent pom.xml specification in `pluginManagement`
   should also be skipped.

Caveat: In the latter case there, we don't yet have support for
verifying that the parent pom.xml file even exists, let alone has the
correct reference for the child.

Addresses a bit of #3405 and #3404.

### CODE REVIEW

File: kythe/go/extractors/config/preprocessor/modifier/modifier_test.go [hunk 1]
```diff
@@ -42,6 +43,7 @@ func TestBuildGradleHasKythe(t *testing.T) {
 		{fileName: "other-build.gradle", javacWrapper: "/tmp/javac-wrapper.sh", wantError: true},
 		// Look at other-build.gradle but use the correct javac wrapper.
 		{fileName: "other-build.gradle", javacWrapper: "/different/javac-wrapper.sh", hasWrapper: true},
+		{fileName: "other-build.gradle", javacWrapper: "/different/javac-wrapper.sh", hasWrapper: true},
```
Comment (danielmoy-google) [inline kythe/go/extractors/config/preprocessor/modifier/modifier_test.go]: I forget what I was doing here, but this pull should be maven only, so axing this

File: kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go [hunk 2]
```diff
@@ -50,10 +56,128 @@ func PreProcessPomXML(pomXMLFile string) error {
 	return f.Close()
 }
 
-func hasCompilerPlugin(doc *etree.Document) bool {
-	for _, p := range doc.FindElements("//project/build/plugins/plugin/artifactId") {
-		if p.Text() == "maven-compiler-plugin" {
+// Returns whether there is a maven-compiler-plugin specified already.
```
Comment (schroederc) [inline kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go]: `// hasCompilerPlugin returns whether...`

File: kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go [hunk 3]
```diff
@@ -50,10 +56,128 @@ func PreProcessPomXML(pomXMLFile string) error {
...
+			// sibling elements.
+			return true, wellformedCompilerPlugin(p)
+		}
+	}
+
+	// If we didn't find anything, we can proceed with modification.
+	return false, nil
+}
+
+// Check to see if the plugin fits the minimum supported versions we've
```
Comment (schroederc) [inline kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go]: `// wellformedCompilerPlugin checks...`

File: kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go [hunk 4]
```diff
@@ -50,10 +56,128 @@ func PreProcessPomXML(pomXMLFile string) error {
...
+	if v, err := javaVersion(k.Text()); err != nil || v.earlierThan(versiontuple{[]int{1, 7}}) {
+		// TODO(#3075): Should test an upper bound here too.
+		return fmt.Errorf("unsupported %s version: %s, need 1.7+", kind, k.Text())
+	}
+	return nil
+}
+
+// Some container that supports semvar (3.7.0) and java versions (1.6) for
+// comparison purposes only.
+type versiontuple struct {
```
Comment (schroederc) [inline kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go]: This can probably just be `type versionTuple []int`.

Comment (danielmoy-google) [reply kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go]: ah yea.  Originally I had this only for semvar and it made sense with the struct, but after refactor to handle both semvar and java version in the same thing, the struct makes very little sense.

Thanks.

File: kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go (hunk 3 above)
Comment (danielmoy-google) [reply kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go]: :+1:

File: kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go (hunk 2 above)
Comment (danielmoy-google) [reply kythe/go/extractors/config/preprocessor/modifier/pom_xml_modifier.go]: :+1:

**Pull request merged**